### DISCUSSION TITLE
typescript definitions

### DISCUSSION THREAD

**Discussion by snuup** [capricorn86/happy-dom#1228] [General]
I was maybe a bit fast with opening this issue: https://github.com/capricorn86/happy-dom/issues/1227

I have a typescript based web framework, that utilises the types in lib.dom.d.ts - the obvious reference. I wonder why happy-dom exposes its interface with own types that are perfectly unrelated with these "public" types. For that reason I used jsdom instead of happy-dom, jsdom exposes the public types. I would like happy-dom to also use the public types, like HTMLElement and so on.

What do you think?

**Comment by snuup** [comment]
Browsing source code on github I find that others struggle exactly like me:

        // JSDOM for document
        // const dom = new JSDOM(bangumiContents);
        // const document = dom.window.document;

        // HappyDOM for document
        // const window = new Window({ url: bangumiURL });
        // const document = window.document as unknown as Document;  // <<<<<<<<<<<<< ugly double cast required
        // // TODO: HappyDOM の型定義を正して、上記の型アノテーションを削除する
        // document.body.innerHTML = bangumiContents;

(https://github.com/xtetsuji/bangumi-js/blob/e67e4893785920bb537ca9db177a85f4998b3896/apps/cli/src/service/bangumi2gcal_service.ts#L5)

**Comment by WebMechanic** [comment]
Same here :-) using WebStorm with JS (not TS) in Vue, Vitest and HappyDOM. I have tsconfig configured for other libs to simplfy my life in the IDE, but HappyDOM somehow interferes. I have not added its types anyplace I am aware of other than for the tests (`packages.json "@happy-dom/global-registrator` and test file `@vitest-environment happy-dom`)

**This is likely more a problem with the IDE, but it'd be helpful if somebody knows how to fix/repair this behaviour.** I don't know enough about TS and typedef resolution to tackle this.

WebStorm apparently finds the type definitions of HappyDOM **overriding it's own built in DOM and Browser types** and then claims things like "Initializer type IElement is not assignable to variable type HTMLElement" for every native method like `document.getElementById`. This is flooding my "problems" panel with false positives.
Adding JSDocs with types doesn't help either. 
```js
/** @type {HTMLElement} */
const eltApp = document.getElementById('app');
```

I didn't realize at first that `IElement` and Co come from HappyDOM. WebStorm also loves to add an automatic import to HappyDOM's node package for the real thing like "HTMLElement", if I add `@type {HTMLElement}` into a comment and press return. 
This often broke my build 'cos I didn't see the import beeing added at the top of the component or script.
I had to disable this feature in the IDE which is unfortunate...

Any ideas and solutions are welcome.
Thanks.

**Comment by capricorn86** [comment]
Happy DOM doesn't apply global types. The Happy DOM types should only be in use for classes directly imported from the 'happy-dom' package. Not when running Happy DOM in the global scope or a VM context with `lib.dom.d.ts`. If that is not the case, it is probably something wrong with the IDE somehow.

Uses Happy DOM types:
```typescript
import { Window, Document } from 'happy-dom';

const window = <Window>new Window();
const document = <Document>window.document;
```

Should not use Happy DOM types (running in a VM context or the global scope):
```typescript
var div = document.createElement('div');
document.body.appendChild(div);
```

**UPDATE:**

To clarify, using types should not be necessary when importing Happy DOM.

```typescript
import { Window, Document } from 'happy-dom';

const window = new Window();
const document = window.document;
```

It seems like Webstorm has a bug, which prioritizes the global types from `lib.dom.d.ts`  before the local types.

You may want to consider to not include `lib.dom.d.ts` for code setting up the `Window` context, or you will have to change name on your variables to names that doesn't collide with the global types (like "window" or "document").

**Comment by capricorn86** [comment]
In the[ latest release](https://github.com/capricorn86/happy-dom/releases/tag/v18.0.0) I fixed so that Happy DOM is compiled using strict typescript configurations. This may fix issues related to having `skipLibCheck` set to "false", which could then cause compilation issues in projects.
